docs/develop-an-agent.md

```diff
@@ -0,0 +1,423 @@
+
+```bash
+
+# URL encode the resource of hub + device + module
+# If we were doing this for the device identity the resource_uri wouldn't include the '/modules/%s'
+resource_uri="$(printf '%s/devices/%s/modules/%s' "$hub_name" "$device_id" "$module_id" | jq -sRr @uri)"
+
+# Using an expiry of 1 day from today
+expiry="$(bc <<< "$(date +%s) + 60 * 60 * 24")"
+
+# Construct the full message to be signed which is a url encoded resource URI concatenated with the expiry and then base64 encoded
+signature_data="$(printf '%s\n%s' "$resource_uri" "$expiry" | base64 -w 0)"
+```
+
+More details on how the IoT Hub service interprets this signed message can be found in the docs on [Security tokens](https://docs.microsoft.com/azure/iot-hub/iot-hub-devguide-security#security-tokens). 
+
+
+#### **Step 3** Get a signed token from the keys service
+
+Send a request to the key service endpoint `/run/aziot/keyd.sock`.
+
+If the connection is refused, go back to Step 1 since IoT Hub device information could have changed.
+
+Otherwise, parse the `signature` from the resposne.
+
+```bash
+# Request the key service sign the message and then parse the 'signature' from the response and url decode it
+signature="$(
+    sudo curl --unix-socket '/run/aziot/keyd.sock' \
+        -X POST -H 'content-type: application/json' --data-raw "$(
+            jq -cn --arg 'keyHandle' "$key_handle" --arg 'message' "$signature_data" '{
+                "keyHandle": $keyHandle,
+                "algorithm": "HMAC-SHA256",
+                "parameters": {
+                    "message": $message,
+                },
+            }'
+        )" 'http://keyd.sock/sign?api-version=2020-09-01' |
+        jq '.signature' -r |
+        sed -e 's|+|%2b|g' -e 's|/|%2f|g' -e 's|=|%3d|g'
+)"
+sas_token="$(printf 'sr=%s&se=%s&sig=%s' "$resource_uri" "$expiry" "$signature")"
+```
+
+#### **Step 4**: Create the full connection string
+
+If the `gatewayHost` retrieved from the identity information is the same as the `hubName`, then the device is configured to connect directly to IoT Hub. Alternatively, if they differ then the device is configured to be behind a gateway.
+
+```bash
+# Note that the 'SharedAccessSignature' appears both in the connection string and in the value for it.
+sas="SharedAccessSignature $sas_token"
+
+if [ $hub_name = $gateway_host ]; then
+    # Connecting directly to Hub.
+    echo "HostName=$hub_name;DeviceId=$device_id;ModuleId=$module_id;SharedAccessSignature=$sas"
+else
+    # In a nested configuration. Include the gateway name in the connection string.
+    echo "HostName=$hub_name;DeviceId=$device_id;ModuleId=$module_id;SharedAccessSignature=$sas;GatewayHost=$gateway_host"
+fi
+
+```
+
+> **Note**
+>
+> Portions of the script above come from the [iothub-get-twin.sh](https://github.com/Azure/iot-identity-service/blob/main/ci/iothub-get-twin.sh) script used in the tests for the identity service.
```

Comment:
```suggestion
> Portions of the script above come from the [iothub-get-twin.sh](https://github.com/Azure/iot-identity-service/blob/main/ci/iothub-get-twin.sh) script used in the tests for the Identity Service.
```

Also, not sure this note is important. Would customers care where this code comes from?

Reply:
Removing it.